### PR TITLE
[WIP] Support relative urls in Component & View

### PR DESCRIPTION
**Pull request by vicb**
**Warning: early proto.**

The hello world example is working with a `templateUrl` (Dart & JS).

Thoughts ?

/cc @mhevery @jakemac53 @tbosch 

TODO:
- [ ] ref #2383

[Review on Reviewable](https://reviewable.io/reviews/angular/angular/2593)

### CODE REVIEW

File: modules/angular2/src/services/url.dart [hunk 1]
```diff
@@ -0,0 +1,11 @@
+library angular2.services.url;
+
+// TODO(vicb): move this to its own module - will be stripped by trsfs
+import 'dart:mirrors';
+
+enum baseUrl {
+  here
+}
+
+String getComponentBaseUrl(Type component)=>
```
Comment (tbosch) [inline modules/angular2/src/services/url.dart]: Why pass in the type and not the annotation?

File: modules/angular2/src/services/url.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import {isPresent, Type, stringify, BaseException} from 'angular2/src/facade/lang';
+import {DirectiveResolver} from 'angular2/src/core/compiler/directive_resolver';
+import {Component} from 'angular2/src/core/annotations_impl/annotations';
+
+
+class BaseUrl {
```
Comment (tbosch) [inline modules/angular2/src/services/url.ts]: I though we would make people provide `module.id`? This would be more general than looking at the stack... Also, think about a minified and concatenated file, which would not contain the module name any more, as reading the stack in JS does not use source maps.

File: modules/angular2/src/core/compiler/compiler.ts [hunk 3]
```diff
@@ -203,9 +201,10 @@ export class Compiler {
     return componentElementBinders;
   }
 
-  private _buildRenderTemplate(component, view, directives): renderApi.ViewDefinition {
-    var componentUrl =
-        this._urlResolver.resolve(this._appUrl, this._componentUrlMapper.getUrl(component));
+  private _buildRenderTemplate(component: Type, view: View,
+                               directives: List<DirectiveBinding>): renderApi.ViewDefinition {
+    var componentUrl = getComponentBaseUrl(component);
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/compiler.ts]: +1 for removing both of these intermediate urls!

File: modules/angular2/src/services/url.ts (hunk 2 above)
Comment (vicb) [inline modules/angular2/src/services/url.ts]: - I don't know how to make `module.id` work, if you can craft a quick ex that works in ng2 context, I'll take a look,
- We don't use the module name but the url from the stack trace (or file path in node ctx, to be implemented)

File: modules/angular2/src/services/url.dart (hunk 1 above)
Comment (vicb) [inline modules/angular2/src/services/url.dart]: for Dart, check the line below.

`baseUrl.here` is only a placeholder in Dart (the enum above) that has to be a compile time constant

File: modules/angular2/src/core/compiler/compiler.ts (hunk 3 above)
Comment (vicb) [inline modules/angular2/src/core/compiler/compiler.ts]: Not sure if this will work for templates compiled server side. In such a case what we can compute is the relative path from the app to the component. 

Then the final url would be app + component relative + templateUrl.

Maybe we can resolve (component relative + templateUrl) at compile time.

File: modules/angular2/src/services/url.dart (hunk 1 above)
Comment (jakemac53) [inline modules/angular2/src/services/url.dart]: We will need to move this into a separate import, along with the import to 'dart:mirrors'. I can probably just do this when I update the transformer though.

File: modules/angular2/src/core/compiler/compiler.ts (hunk 3 above)
Comment (mhevery) [inline modules/angular2/src/core/compiler/compiler.ts]: I think we should resolve the URLs to absolute at compile or inline the templates so there are no URLs.

File: modules/angular2/src/core/compiler/template_resolver.ts [hunk 4]
```diff
@@ -22,7 +22,7 @@ export class TemplateResolver {
     return view;
   }
 
-  _resolve(component: Type) {
+  _resolve(component: Type): View {
```
Comment (mhevery) [inline modules/angular2/src/core/compiler/template_resolver.ts]: Thanks for adding types. The more the easier the docs generation will be.

File: modules/angular2/src/render/api.ts [hunk 5]
```diff
@@ -143,10 +143,11 @@ export class DirectiveMetadata {
   hostProperties: Map<string, string>;
   hostAttributes: Map<string, string>;
   hostActions: Map<string, string>;
+  baseUrl: any;
```
Comment (mhevery) [inline modules/angular2/src/render/api.ts]: `baseUrl: string`?

File: modules/angular2/src/services/url.dart (hunk 1 above)
Comment (mhevery) [inline modules/angular2/src/services/url.dart]: @vicb please address @jakemac53 comment. We can't have mirrors. For any mirrors code we must be able to replace it with a different code which is mirror independent during transformers.

File: modules/angular2/src/services/url.ts (hunk 2 above)
Comment (mhevery) [inline modules/angular2/src/services/url.ts]: I like the stack trick, its cleaner. 

@tbosch I don't think this is an issue for bundling, since there we will inline content / expand the URLs as a build step, so it will not be an issue.

File: modules/angular2/src/services/url.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  // ugly, chrome specific
+  static re: RegExp = /(?:\((.*?):\d+:\d+)|(?:at ([^(]+):\d+:\d+)/;
```
Comment (mhevery) [inline modules/angular2/src/services/url.ts]: Maybe grab stack traces for all major browsers and craft a test which shows that we can properly process all of them.

File: modules/angular2/src/services/url.dart (hunk 1 above)
Comment (mhevery) [inline modules/angular2/src/services/url.dart]: Can't we use the same stacktrace trick in Dart as well? We don' t need the mirrors at all then?

File: modules/examples/src/hello_world/index_common.ts [hunk 7]
```diff
@@ -29,14 +31,14 @@ class RedDec {
   selector: 'hello-app',
   // These are services that would be created if a class in the component's
   // template tries to inject them.
-  appInjector: [GreetingService]
+  appInjector: [GreetingService],
+  baseUrl: baseUrl.here
```
Comment (mhevery) [inline modules/examples/src/hello_world/index_common.ts]: Can't the annotation extract the stack trace automatically? We don't need to pass it explicitly. In the annotation constructor we can just throw and extract (if no baseUrl is specified). I think this will make the feature truly awesome!

**Comment by mhevery** [comment]
Can you look at https://github.com/angular/angular/pull/2593#discussion_r32665828 first?

File: modules/angular2/src/services/url.dart (hunk 1 above)
Comment (jakemac53) [inline modules/angular2/src/services/url.dart]: The mirrors should be easy to replace, since we will be inlining these anyways. We just need to replace the import which provides this function with a new import that throws an unimplemented error.

Comment (jakemac53) [inline modules/angular2/src/services/url.dart]: @mhevery if we stick with mirrors, do you want the transformer to be updated in the same cl? I was planning on just following up with a cl to update the transformer, but that would temporarily leave us in a state with mirrors.

Comment (vicb) [inline modules/angular2/src/services/url.dart]: We have discussed moving mirror to a specific file with @jakemac53. This is not done yet because this is an early proto.

@jakemac53 maybe you can create the code in the PR that I'll cherry pick here (I can adapt the src/target import names at the same time)

Comment (jakemac53) [inline modules/angular2/src/services/url.dart]: That sounds good, I probably won't be able to do it until tomorrow though.

File: modules/angular2/src/core/compiler/compiler.ts (hunk 3 above)
Comment (vicb) [inline modules/angular2/src/core/compiler/compiler.ts]: > I think we should resolve the URLs to absolute at compile or inline the templates so there are no URLs.

Can we resolve the URL at compile time ? We would need to hardcode the base URLs which IMO is not nice.

Long term we should inline anyway but I think this is out of scope for this PR, right ?

File: modules/angular2/src/render/api.ts (hunk 5 above)
Comment (vicb) [inline modules/angular2/src/render/api.ts]: `baseUrl: baseUrl` in Dart where it is an enum

File: modules/angular2/src/services/url.ts (hunk 6 above)
Comment (vicb) [inline modules/angular2/src/services/url.ts]: I've spotted some existing packages for that, I need to figure out the best one.

File: modules/examples/src/hello_world/index_common.ts (hunk 7 above)
Comment (vicb) [inline modules/examples/src/hello_world/index_common.ts]: Great idea !!!

File: modules/angular2/src/services/url.dart (hunk 1 above)
Comment (vicb) [inline modules/angular2/src/services/url.dart]: No problem, I'll work on it tomorrow morning and then I'll be off until Monday.

**Comment by vicb** [comment]
> Can you look at #2593 (comment) first?

Yep, seen it. It's a GREAT idea. I should also be able to remove the `enum` hack for Dart as well with it.

File: modules/angular2/src/services/url.dart (hunk 1 above)
Comment (vicb) [inline modules/angular2/src/services/url.dart]: @jakemac53 What are your thoughts about stack traces vs mirrors/transformers for Dart ? 

`print(new Trace.current().frames.map((f) => f.uri).toList());` (Trace is from the pub package stack_trace) seems to work fine.

However I feel like mirrors/transformers could be more solid... (as it does not depend on the underlying browser)

Comment (jakemac53) [inline modules/angular2/src/services/url.dart]: It seems to me like the mirrors version would be less fragile?

Comment (vicb) [inline modules/angular2/src/services/url.dart]: +1

quick question: what happens if you can not get a `package://` url as of today (the component is located in one of your app folder), is this currently supported ?

Comment (jakemac53) [inline modules/angular2/src/services/url.dart]: Not sure exactly what you mean, but `package:` urls and relative urls should be supported. Are you talking about pointing to a file in say `web` from `test`?

**Comment by tbosch** [comment]
Please consider Safari and SystemJS:
SystemJS uses `eval` and gives the eval a filename by adding a `//#sourceURL`. However, when an Error is thrown in Safari, these names are not reflected in the stack, so this method won't work on Safari.

**Comment by naomiblack** [comment]
@vicb #1694 is now closed. I think this unblocks the remaining work?

**Comment by vicb** [comment]
I don't think it has ever been blocked but it takes time as it is non trivial. Working on this today an next week.

**Comment by naomiblack** [comment]
@vicb can you provide an update? is this still likely to land by july 6?

**Comment by tbosch** [comment]
@vicb Did you close this on purpose?

**Comment by vicb** [comment]
@tbosch this has been closed because it uses stack traces which does not work cross browsers. I'll detail all of this in a doc.
